Compare checksum for repodata against the downloaded rpm file

shivania2 opened the pull request:
tdnf when downloading packages (or using from a repo accessible as a file), does not check for the checksum which is provided in the metadata. Recently there is a issue caught https://bugzilla.eng.vmware.com/show_bug.cgi?id=3237405(Fixed by checking the size), a proper check would have caught this issue earlier.

oliverkurth commented on client/rpmtrans.c:
```diff
@@ -898,6 +899,16 @@ TDNFTransAddInstallPkg(
+    dwError = TDNFGetSHA256DigestForFile(pszFilePath, digest_from_file);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (memcmp(digest_from_file, pInfo->pszChecksum, SHA256_DIGEST_LENGTH))
```
The reason this fails is that you are comparing a string in hexadecimal format (`pInfo->pszChecksum`) with a binary checksum (`digest_from_file`). You need to convert the hex string to binary and then compare. See how it's done in `plugins/metalink/utils.c`:
```
            TDNFCheckHexDigest(currHashInfo->value, hash_ops[currHashType].length)) {
            dwError = TDNFChecksumFromHexDigest(currHashInfo->value, digest);
            BAIL_ON_TDNF_ERROR(dwError);

            dwError = TDNFCheckHash(pszFile, digest, hash_Type);
```
Here the function `TDNFChecksumFromHexDigest` converts from hex string to binary.

oliverkurth commented on client/utils.c:
```diff
@@ -330,6 +330,104 @@ TDNFGetFileSize(
     goto cleanup;
 }
 
+uint32_t
```
I see you have copied this from `plugins/metalink/utils.c` and slightly modified.

Instead please just move those functions (copy here and remove in `plugins/metalink/utils.c`) , and make the metalink plugin use those functions from here (you may have to move them to `common/utils.c` instead of `client/utils.c`, not sure).

That won't work with these modifications because you degeneralized this to just use sha256.

oliverkurth commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,58 @@ SolvGetPackageId(
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pszTemp = solvable_lookup_checksum(pSolv, SOLVABLE_CHECKSUM, &checksumtype);
```
Looks like `libsolv` has a function to get the checksum in binary: https://github.com/openSUSE/libsolv/blob/2fc0cb31a2010d70f3f9197189730400edbd6ef4/src/solvable.h#L76

oliverkurth commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,58 @@ SolvGetPackageId(
+    const char* pszTemp = NULL;
+    char* pszChecksum = NULL;
+    Solvable *pSolv = NULL;
+    int checksumtype= REPOKEY_TYPE_SHA256;
```
Can you find out if repo data can use other types than sha256?

oliverkurth replied:
Look at https://github.com/rpm-software-management/createrepo_c , it looks like it supports different hash types:

checksum | Checksum type to use | md5, sha, sha1, sha224, sha256, sha384, sha512 | sha256

oliverkurth reviewed (changes requested):
I haven't found out why, but somehow this breaks removing packages:
```
# ./bin/tdnf install lsof

Installing:
libtirpc                               x86_64                       1.3.3-2.ph5                            photon-updates               194.10k                          92.17k
lsof                                   x86_64                       4.96.4-1.ph5                           photon-updates               207.83k                         120.93k

Total installed size: 401.92k
Total download size: 213.10k
Is this ok [y/N]: y
libtirpc                                 94380 100%
lsof                                    123832 100%
Testing transaction
Running transaction
Installing/Updating: libtirpc-1.3.3-2.ph5.x86_64
Installing/Updating: lsof-4.96.4-1.ph5.x86_64

# ./bin/tdnf remove lsof
Nothing to do.
```

oliverkurth commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+error:
+    if (!IsNullOrEmptyString(filename))
+    {
+        pr_err("Error: Validating metalink (%s) FAILED (digest mismatch)\n", filename);
```
You need to remove references to "metalink" since the code isn't specific to it any more.

oliverkurth commented on client/packageutils.c:
```diff
@@ -1148,6 +1148,12 @@ TDNFPopulatePkgInfos(
                       &pPkgInfo->pszLocation);
         BAIL_ON_TDNF_ERROR(dwError);
 
+        dwError = SolvGetPkgChecksumFromId(
```
This shouldn't fail when there is no checksum available. This is why removing packages fails - it tries to get the checksum from an installed package, which isn't there.

We should check the error for `ERROR_TDNF_NO_DATA`, and in that case ignore the error. There may also be repositories without any checksum data (?), and we shouldn't fail with those (or maybe make it conditional with an option).

Later, when we want to check the checksum we can check for `pPkgInfo->pszChecksum == NULL`.

oliverkurth commented on client/rpmtrans.c:
```diff
@@ -898,6 +900,20 @@ TDNFTransAddInstallPkg(
+
+        if (memcmp(digest_from_file, pInfo->pszChecksum, hash->length))
+        {
+            pr_err("rpm file (%s) checksum (%s) does not match repodata checksum (%s)\n", pszFilePath, digest_from_file, pInfo->pszChecksum);
```
You are printing binary data here in the error message. That's why we get failures in the github test run like this:
```
        stdout = out.decode().strip()
>       stderr = err.decode().strip()
E       UnicodeDecodeError: 'utf-8' codec can't decode byte 0xcf in position 137: invalid continuation byte
```

But I don't understand why this error message appears in the first place - the checksum should match. And it works when I test this locally.

oliverkurth commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,58 @@ SolvGetPackageId(
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFAllocateString((char *)pszTemp, (char **)&pszChecksum);
```
Ah, this is wrong: `solvable_lookup_bin_checksum()` is returning binary data, not a string. You need to figure out the length of the data and use `TDNDAllocateMemory()`, then copy the data.

oliverkurth commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,61 @@ SolvGetPackageId(
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pszTemp_size = strlen((const char*)pszTemp) + 1;
```
Again, this is binary data, not a `nul` terminated string. You cannot use `strlen()` for it.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+    fd = open(filename, O_RDONLY);
+    if (fd < 0)
+    {
+        pr_err("Metalink: validating (%s) FAILED\n", filename);
```
Is it only related to metalink if not this error message looks confusing.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        /* MD5 is not approved in FIPS mode. So, overrriding
+           the dwError to show the right error to the user */
+#if defined(OPENSSL_VERSION_MAJOR) && (OPENSSL_VERSION_MAJOR >= 3)
```
Looks quite open can put a macro function and call it.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    while ((length = read(fd, buf, (sizeof(buf)-1))) > 0)
```
instead of sizeof call you can do BUFSIZ - 1 as we know the size of buf.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+
+    if (length == -1)
+    {
+        pr_err("Metalink: validating (%s) FAILED\n", filename);
```
See the above comment is it only related to metalink or not.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+       (digest_length <= 0))
+    {
+        return 0;
+    }
```
\n after this for better view.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+    val = strtoul(buf, NULL, 16);
+    if(errno)
+    {
+        pr_err("Error: strtoul call failed\n");
```
This is correct way to log error msg. Wherever we used pr_err i was expecting like that Error: <error msg> that will help while debuging failures.

ntsbtz commented on common/utils.c:
```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+        dwError = errno;
+        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+    }
+    *uintValue = (unsigned char)(val&0xff);
```
Readablity purpose 
*uintValue = (unsigned char)(val&0xff) --> *uintValue = (unsigned char)(val & 0xff);

ntsbtz commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,60 @@ SolvGetPackageId(
+SolvGetPkgChecksumFromId(
+    PSolvSack pSack,
+    uint32_t dwPkgId,
+    int checksumtype,
```
Maintain the variable naming generic through out the code
int checksumtype --> int checksumType
int checksumlen --> int checksumLen

ntsbtz commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,60 @@ SolvGetPackageId(
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    memcpy(pszChecksum, pszTemp, checksumlen);
+    *ppszChecksum = pszChecksum;
```
Outside this function there is a memory leak not freeing the ppszChecksum memory .
once you return from this function while cleanup TDNFFreePackageInfo is getting called need to add free for ppszChecksum.

oliverkurth replied:
Agree with Nitesh, you need to clean in `TDNFFreePackageInfoContents()` in `common/utils.c`.

ntsbtz commented on client/packageutils.c:
```diff
@@ -1148,6 +1148,19 @@ TDNFPopulatePkgInfos(
+                      pSack,
+                      dwPkgId,
+                      REPOKEY_TYPE_SHA256,
+                      SIZEOF_SHA256,
```
This function only support for SHA256 ?

ntsbtz replied:
> This function only support for SHA256 ?

If this is not specific for SHA256  then we can consider to check other checksum as well.
But if the expectation is only for SHA256 then its fine.

oliverkurth replied:
I commented before that repodata can have other checksum types.

ntsbtz reviewed (changes requested):
There is some review comments please address them before merging.

ntsbtz reviewed (approved):
Thanks for taking care of review comments.
LGTM !!! apart from SHA256 check is little confusing for me.

oliverkurth commented on solv/tdnfpackage.c:
```diff
@@ -820,6 +820,60 @@ SolvGetPackageId(
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFAllocateMemory(checksumLen, sizeof(unsigned char), (void **)&pszChecksum);
```
Okay good, this is better now.

Please change the variable name `pszChecksum` to something like `pChecksum` or `pbChecksum`. The `psz` prefix means "pointer to a zero terminated string", which this is not. You can google for "hungarian notation".

Also for `ppszChecksum`.

oliverkurth reviewed (changes requested):
Please use valgrind to check for memory leaks.

Also, run this with coverity once.

oliverkurth commented on client/packageutils.c:
```diff
@@ -1094,6 +1094,7 @@ TDNFPopulatePkgInfos(
     Id dwPkgId = 0;
     PTDNF_PKG_INFO pPkgInfos = NULL;
     PTDNF_PKG_INFO pPkgInfo = NULL;
+    int ChecksumType = 0;
```
use `nChecksumType`.

oliverkurth commented on include/tdnftypes.h:
```diff
@@ -153,6 +153,7 @@ typedef struct _TDNF_PKG_INFO
     uint32_t dwEpoch;
     uint32_t dwInstallSizeBytes;
     uint32_t dwDownloadSizeBytes;
+    int ichecksumType;
```
use `nChecksumType`

oliverkurth reviewed (changes requested):
This is looking really good now!

Just a few nits, and please write a test.

I tested this also:
* I created a repo with two packages and changed the checksum in the metadata for one of them, it failed to install as expected with the 1528 error message
* I tested the repo with sha512 (`-s sha512` option to `createrepo`), and that worked too
* I also tested with `valgrind`, it found no memory leaks.

Please write a test. You can manipulate the checksum in the metadata data this way:
* create a repo with a few packages with `createrepo`
* the metadata with the checksums are in the directory `repodata` in files ending with  `.primary.xml.gz`
* you need to unzip the file with `gunzip`, then change the checksum for a package
* compress the file again with the same name
* try to install, it should fail.

If the repo is for example in `/repo`, you can test an install with:
`tdnf --repofrompath=test,/repo --repoid=test install lsof`

See for example `pytests/tests/test_repofrompath.py`

shivania2 commented:
> This is looking really good now!
> 
> Just a few nits, and please write a test.
> 
> I tested this also:
> 
> * I created a repo with two packages and changed the checksum in the metadata for one of them, it failed to install as expected with the 1528 error message
> * I tested the repo with sha512 (`-s sha512` option to `createrepo`), and that worked too
> * I also tested with `valgrind`, it found no memory leaks.
> 
> Please write a test. You can manipulate the checksum in the metadata data this way:
> 
> * create a repo with a few packages with `createrepo`
> * the metadata with the checksums are in the directory `repodata` in files ending with  `.primary.xml.gz`
> * you need to unzip the file with `gunzip`, then change the checksum for a package
> * compress the file again with the same name
> * try to install, it should fail.
> 
> If the repo is for example in `/repo`, you can test an install with: `tdnf --repofrompath=test,/repo --repoid=test install lsof`
> 
> See for example `pytests/tests/test_repofrompath.py`

Able to create sha512 repo, modified the checksum and repacked the files ending with  `.primary.xml.gz`. But package install is failed with command "tdnf --repofrompath=test,/repo --repoid=test install lsof"

oliverkurth commented on pytests/tests/test_install.py:
```diff
@@ -22,7 +26,46 @@ def setup_test(utils):
+    if os.path.isdir(WORKDIR):
+        shutil.rmtree(WORKDIR)
+
+def create_repo(utils):
```
Please move all the checksum tests into a new `test_checksum.py` (or similar) file.

oliverkurth commented on pytests/tests/test_install.py:
```diff
@@ -22,7 +26,46 @@ def setup_test(utils):
+    assert os.path.isdir(synced_dir)
+    assert os.path.isdir(os.path.join(synced_dir, 'repodata'))
+    assert os.path.isfile(os.path.join(synced_dir, 'repodata', 'repomd.xml'))
+
```
Nit: needs two empty lines between functions (`flake8` will call this out)

oliverkurth commented on pytests/tests/test_install.py:
```diff
@@ -22,7 +26,46 @@ def setup_test(utils):
+                    # Compress the updated XML data again with the same name
+                    with gzip.open(file_path, 'wt', encoding='utf-8') as f:
+                        f.write(ET.tostring(root, encoding='utf-8').decode())
 
```
Also here and other places.

oliverkurth reviewed (changes requested):
Thanks for adding tests - now you need to find out why they are failing ;-)

oliverkurth commented on pytests/tests/test_checksum.py:
```diff
@@ -0,0 +1,132 @@
+        elif filename.startswith(new_pkg):
+            suffix = find_suffix(filename, new_pkg)
+
+            if not os.path.exists(orig_path):
```
This won't work if you find `new_pkg` first - `orig_path` will not be set (it will still be '')

oliverkurth commented on pytests/tests/test_checksum.py:
```diff
@@ -0,0 +1,132 @@
+            if not os.path.exists(orig_path):
+                orig_filename = orig_pkg + '-1.0.1-2.x86_64.rpm'
+                orig_path = os.path.join(workdir, orig_filename)
+                file = open(orig_path, 'w')
+                file.close()
```
What is the purpose of opening then closing the file?

oliverkurth commented on pytests/tests/test_checksum.py:
```diff
@@ -0,0 +1,132 @@
+    reponame = 'photon-test-sha512'
+    synced_dir = os.path.join(workdir, reponame)
+
+    enable_and_create_repo(utils)
```
Since you are not modifying the repo for this test, you can just use `photon-test-sha512` directly.

oliverkurth commented on pytests/tests/test_checksum.py:
```diff
@@ -0,0 +1,132 @@
+        return None
+
+
+def modify_rpm(workdir, orig_pkg, new_pkg):
```
I am not sure what this function is supposed to do.

If it's to copy `orig_pkg` to `new_pkg`, this can be done much simpler.

oliverkurth commented on pytests/tests/test_checksum.py:
```diff
@@ -0,0 +1,117 @@
+    orig_path = ''
+
+    for filename in os.listdir(workdir):
+        if filename.startswith(orig_pkg):
```
If we find `new_pkg` before `orig_pkg`, this won't work because it wont't set `orig_path`, right?

This function is very confusing. Please rewrite it with a simpler version.

oliverkurth reviewed (changes requested):
The functional changes look good to me, but you need to make changes in the test code. I don't think it works as intended.
